Feature: Resolve rebase conflicts

Murmele opened the pull request:
- [x] Unittests
- [x] All TODOs removed
- [x] ~~Toolbar buttons for aborting and continuing the rebase~~  Moved next to the commit button
- [x] Merge conflicts and mergetool are enabled!

Murmele commented:
@exactly-one-kas I implemented it rawly now. What do you think about the toolbuttons or do you see a better location?

exactly-one-kas commented:
I think that aborting and continuing rebases should be handled like merges; otherwise the merge operations would also habe to be placed on the toolbar for consistency

![image](https://user-images.githubusercontent.com/16035731/177027055-d19785ca-2a13-41c8-8d15-de54d9e9a290.png)
![image](https://user-images.githubusercontent.com/16035731/177027061-9e25f93f-d6d8-4f88-97c6-dbdc9d9c3326.png)

Murmele commented:
> I think that aborting and continuing rebases should be handled like merges; otherwise the merge operations would also habe to be placed on the toolbar for consistency
> 
> ![image](https://user-images.githubusercontent.com/16035731/177027055-d19785ca-2a13-41c8-8d15-de54d9e9a290.png) ![image](https://user-images.githubusercontent.com/16035731/177027061-9e25f93f-d6d8-4f88-97c6-dbdc9d9c3326.png)

The second image is good. But maybe we can then also move the merge out. I think it is quite hidden. I did not know ^^

Murmele commented:
I moved the rebase buttons.
I added also the merge abort button there or should I let it in the menubar only, or should I delete it from the menubar then?

Murmele commented:
@exactly-one-kas do you understand why the unittest fails with that exact repository? For some reason I have a nullptr exception 

![grafik](https://user-images.githubusercontent.com/10099533/177107064-701f4fbf-6423-4e15-bbff-7284c92de7a6.png)

exactly-one-kas commented:
Does the unit test only fail with that exact repository? Considering the stack strace it seems to me that it should fail with any repository.

I'd guess that this happens because the unit tests are run with an offscreen / dummy QT backend that doesn't create actual windows and therefore has no window handles.

Try the following:
- Rename `w_handler` to `w_handle`
- Use `int size = kSize * (w_handle ? w_handle->devicePixelRation() : 1);`

exactly-one-kas commented:
> I moved the rebase buttons. I added also the merge abort button there or should I let it in the menubar only, or should I delete it from the menubar then?

I'd say that all buttons regarding rebasing and merging should be with the commit button (after all, the "commit merge" button *is* the merge button), so that new users have an easier time finding these options. Personally, I think that this location is the more logical one than the toolbar or menu bar

However, "abort merge" and "abort rebase" should also be present in the "branch" menu as that's where users currently expect them to be. We should keep them there for a while and phase them out for v2 / the next major redesign

Murmele commented:
> 

Yes exactly it fails only with this repository, If I replace the zip filename by another in that location it runs through.

Murmele commented:
> > I moved the rebase buttons. I added also the merge abort button there or should I let it in the menubar only, or should I delete it from the menubar then?
> 
> I'd say that all buttons regarding rebasing and merging should be with the commit button (after all, the "commit merge" button _is_ the merge button), so that new users have an easier time finding these options. Personally, I think that this location is the more logical one than the toolbar or menu bar
> 
> However, "abort merge" and "abort rebase" should also be present in the "branch" menu as that's where users currently expect them to be. We should keep them there for a while and phase them out for v2 / the next major redesign

Good idea, I will mark it so we don't forgett to remove it at some point

Murmele commented:
This is how the stacktraces look like

![GittyupRepoLoad drawio](https://user-images.githubusercontent.com/10099533/177943357-d2e49c6a-10e6-4953-94d0-2218d7c33dc8.png)

What I have seen is that for rebaseConflicts Repository, the number of commits is not zero as for the other 
![grafik](https://user-images.githubusercontent.com/10099533/177943697-61299bb7-d0f2-4ebd-8156-828db78b0e7a.png)

Murmele commented on src/dialogs/RebaseConflictDialog.cpp:
```diff
@@ -24,18 +24,16 @@ RebaseConflictDialog::RebaseConflictDialog(QWidget *parent) : QDialog(parent) {
           });
 
-  connect(buttons->addButton(tr("Keep rebase"), QDialogButtonBox::AcceptRole),
+  connect(buttons->addButton(tr("Fix merge conflicts"), QDialogButtonBox::AcceptRole),
```
Continue sounds fine. I will fix it. merge conflicts sounds also not completely correct for me

Murmele replied:
done in 7951323e3217ca12538ae6a74a292e35c3fa6fdb

Murmele commented on CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ if(GIT_FOUND)
             --format=%h HEAD
     OUTPUT_VARIABLE GITTYUP_BUILD_REVISION
     OUTPUT_STRIP_TRAILING_WHITESPACE)
+    add_compile_definitions(GIT_EXECUTABLE="${GIT_EXECUTABLE}") # To be able to access git cli for testing purpose
```
you are right an can just copy it there. Don't know why I thought it is only available from here

Murmele replied:
fixed in 2e897e166eafe4570d797fb15595ece3916209a9

Murmele commented on src/git/Rebase.h:
```diff
@@ -12,8 +12,12 @@
+
 struct git_rebase;
 struct git_repository;
+//struct git_rebase_operation;
```
thanks :)

Murmele commented on src/git/Repository.cpp:
```diff
@@ -830,13 +830,98 @@ bool Repository::merge(const AnnotatedCommit &mergeHead) {
+
+    if (r.finish())
+        emit d->notifier->rebaseFinished(r, parent);
+    //else
```
For now it is a placeholder. I have to check out what do do there in case the finish fails

Murmele replied:
added TODO: so I don't forgett

Murmele replied:
7951323e3217ca12538ae6a74a292e35c3fa6fdb

Murmele commented:
> Watching `.git/HEAD` would be an option

hmm yes. So we need another application always looking there. I think this goes beyond this scope.

exactly-one-kas commented:
> hmm yes. So we need another application always looking there. I think this goes beyond this scope.

Why would this need to be in another application? Change detection happens in the same process from another thread (see `src/watcher/RepositoryWatcher.cpp`)

Murmele commented:
> > hmm yes. So we need another application always looking there. I think this goes beyond this scope.
> 
> Why would this need to be in another application? Change detection happens in the same process from another thread (see `src/watcher/RepositoryWatcher.cpp`)

You are right sorry. This will work so we will react on .git/HEAD changes and then we are doing a refresh? Would you like to implement ^^ I am completely full next weeks. Maybe I can go on a little bit with the libgit2 update ....

Murmele commented:
@exactly-one-kas can we merge this? I think implementing watching .git/HEAD can be done separately because it affects all commands executed in command line not only rebase

exactly-one-kas commented:
Yep, works for me
